py/server/deephaven/_udf.py

```diff
@@ -536,9 +536,10 @@ def _udf_decorator(encoded_arg_types: str, for_vectorization: bool):
         """
         arg_conv_needed = p_sig.prepare_auto_arg_conv(encoded_arg_types)
         p_sig.ret_annotation.setup_return_converter()
+        ret_converter = p_sig.ret_annotation.ret_converter
 
         if not for_vectorization:
-            if not arg_conv_needed and p_sig.ret_annotation.encoded_type == "O":
+            if not arg_conv_needed and not ret_converter and not return_array:
                 return fn
 
             def _wrapper(*args, **kwargs):
```

Comment:
This wrapper is still quite branchy.  If the branching was pulled outside of the wrapper, I think we end up with better performance.